File: guessing-game-bdd-atomic/src/test/java/tw/waterballsa/bddworkshop/stepdefs/given/Given_A_Greeting.java
package tw.waterballsa.bddworkshop.stepdefs.given;

import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Given;
import org.springframework.beans.factory.annotation.Autowired;
import tw.waterballsa.bddworkshop.entity.Greeting;
import tw.waterballsa.bddworkshop.repository.GreetingRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public class Given_A_Greeting {

    @Autowired
    private GreetingRepository greetingRepository;

    @Given("a greeting, with:")
    public void invoke(DataTable dataTable) throws Exception {
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        
        for (Map<String, String> row : rows) {
            String id = row.get("id");
            String content = row.get("content");
            String name = row.get("name");
            String createdAtStr = row.get("createdAt");
            
            // Handle special <now_ts> placeholder
            LocalDateTime createdAt;
            if ("<now_ts>".equals(createdAtStr)) {
                createdAt = LocalDateTime.now();
            } else {
                // Parse the timestamp string if it's not the placeholder
                createdAt = LocalDateTime.parse(createdAtStr);
            }
            
            Greeting greeting = new Greeting(id, content, name, createdAt);
            
            greetingRepository.save(greeting);
        }
    }
}

File: guessing-game-bdd-atomic/src/test/stepdefs/given/Given_A_Greeting.ts
import { Given } from '@cucumber/cucumber';
import { DataTable } from '@cucumber/cucumber';
import { INestApplication } from '@nestjs/common';
import { Repository } from 'typeorm';
import { Greeting } from '../../../main/greeting/entities/greeting.entity';

export class Given_A_Greeting {
    constructor(
        private readonly app: INestApplication
    ) {}

    @Given('a greeting, with:')
    public async invoke(dataTable: DataTable): Promise<void> {
        const greetingRepository = this.app.get<Repository<Greeting>>('GreetingRepository');
        const rows = dataTable.hashes();

        for (const row of rows) {
            const { id, content, name, createdAt: createdAtStr } = row;

            // Handle special <now_ts> placeholder
            let createdAt: Date;
            if (createdAtStr === '<now_ts>') {
                createdAt = new Date();
            } else {
                // Parse the timestamp string if it's not the placeholder
                createdAt = new Date(createdAtStr);
            }

            const greeting = new Greeting(id, content, name, createdAt);
            await greetingRepository.save(greeting);
        }
    }
}
